### PR TITLE
HAAR-2176: Added migrate endpoint for client credentials.

### PR DESCRIPTION
**Pull request by ravmoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/MigrationClientService.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  @Transactional
+  fun addClient(migrationClientRequest: MigrationClientRequest) {
+    val clientList = clientIdService.findByBaseClientId(migrationClientRequest.clientId!!)
+    if (clientList.isNotEmpty()) {
+      throw ClientAlreadyExistsException(migrationClientRequest.clientId)
+    }
+
+    var client = conversionService.convert(migrationClientRequest, Client::class.java)
+    client!!.clientSecret = migrationClientRequest.clientSecret
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/MigrationClientService.kt:40]: Added 'client secret' to converter

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/MigrationClientService.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  @Transactional
+  fun addClient(migrationClientRequest: MigrationClientRequest) {
+    val clientList = clientIdService.findByBaseClientId(migrationClientRequest.clientId!!)
+    if (clientList.isNotEmpty()) {
+      throw ClientAlreadyExistsException(migrationClientRequest.clientId)
+    }
+    var client = conversionService.convert(migrationClientRequest, Client::class.java)
+
+    client = client?.let { clientRepository.save(it) }
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/MigrationClientService.kt:33]: Do we need the ? operator here? The client returned from the conversion service will never be null?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/integration/MigrationControllerIntTest.kt [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        null,
+      )
+
+      clientRepository.delete(client)
+      clientConfigRepository.delete(clientConfig)
+      authorizationConsentRepository.delete(authorizationConsent)
+    }
+
+    @Test
+    fun `register incomplete client`() {
```
Comment (AlTurner-MOJ) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/integration/MigrationControllerIntTest.kt:264]: rename this test case **_migrate incomplete client_**

File: src/test/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/integration/MigrationControllerIntTest.kt [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        "AuthorizationServerDetailsMigrate",
+        mapOf("username" to "AUTH_ADM", "clientId" to "testp"),
+        null,
+      )
+
+      clientRepository.delete(client)
+    }
+
+    @Test
+    fun `omit mandatory fields gives bad request response`() {
```
Comment (AlTurner-MOJ) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/integration/MigrationControllerIntTest.kt:302]: I don't think we need this test case here - we should make clientId a mandatory field on the request and remove the validation annotations

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/MigrationClientService.kt (hunk 2 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/MigrationClientService.kt]: agreed , removed now

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    clientDetails: MigrationClientRequest,
+  ): ResponseEntity<Any> {
+    val registrationResponse = migrationClientService.addClient(clientDetails)
+    val telemetryMap = mapOf("username" to authenticationFacade.currentUsername!!, "clientId" to clientDetails.clientId!!)
+    telemetryClient.trackEvent("AuthorizationServerDetailsMigrate", telemetryMap)
+    return ResponseEntity.ok(registrationResponse)
+  }
+}
+
+class MigrationClientRequest(
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt:39]: agreed and added

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/MigrationClientService.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/service/MigrationClientService.kt]: added

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  val clientId: String?,
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt:40]: You can make clientId mandatory - plus remove !! wherever it is referenced

File: src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+) {
+
+  @PostMapping("migrate-client")
+  @ResponseStatus(HttpStatus.OK)
+  @PreAuthorize("hasRole('ROLE_OAUTH_ADMIN')")
+  fun addClient(
+    @Valid @RequestBody
+    clientDetails: MigrationClientRequest,
+  ): ResponseEntity<Any> {
+    val registrationResponse = migrationClientService.addClient(clientDetails)
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt:32]: Question - do we need to capture the response here? The service does not return anything (think default return type in Kotlin is Unit?). Would it be clearer to define the end point as not returning anything?

Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/authorizationserver/resource/MigrationController.kt:32]: removed response, it was required at all as it already have annotation  @ResponseStatus(HttpStatus.OK)

**Pull request merged**